src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,211 @@ namespace NServiceBus
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
+            public CorrelationId CorrelationId { get; private set; }
+            public IContainSagaData ContainSagaData { get; }
+
+            public Entry(IContainSagaData sagaData, CorrelationId correlationId)
             {
-                VersionedSagaEntity value;
-                if (data.TryRemove(sagaData.Id, out value))
-                {
-                    object lockToken;
-                    lockers.TryRemove(value.LockTokenKey, out lockToken);
-                }
-            });
-            return TaskEx.CompletedTask;
+                CorrelationId = correlationId;
+                ContainSagaData = DeepClone(sagaData);
+            }
+
+            static IContainSagaData DeepClone(IContainSagaData source)
+            {
+                var json = serializer.SerializeObject(source);
+                return (IContainSagaData)serializer.DeserializeObject(json, source.GetType());
+            }
+
+            public Entry UpdateTo(IContainSagaData sagaData)
+            {
+                return new Entry(sagaData, CorrelationId);
+            }
         }
 
-        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task Complete(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            Guard.AgainstNull(nameof(propertyValue), propertyValue);
-
-            foreach (var entity in data.Values)
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
             {
-                if (!(entity.SagaData is TSagaData))
+                var entry = GetEntry(sagaData, context);
+
+                if (_sagas.TryRemoveConditionally(sagaData.Id, entry) == false)
                 {
-                    continue;
+                    throw new Exception("Saga can't be completed as it was updated by another process.");
                 }
 
-                var prop = typeof(TSagaData).GetProperty(propertyName);
-                if (prop == null)
+                // saga removed
+                // clean the index
+                if (Equals(entry.CorrelationId, NoCorrelationId) == false)
                 {
-                    continue;
+                    _byCorrelationId.TryRemoveConditionally(entry.CorrelationId, sagaData.Id);
                 }
-                var existingValue = prop.GetValue(entity.SagaData);
+            });
+
+            return Task.FromResult(0);
+        }
+
+        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context)
+            where TSagaData : IContainSagaData
+        {
+            Entry value;
+
+            if (_sagas.TryGetValue(sagaId, out value))
+            {
+                context.Set(sagaId.ToString(), value);
```

Comment:
Why wouldn't it hold up? Could you elaborate?
For every `Update`/`Complete` called on persister the saga must be retrieved with `Get` first.

You can't use a constant key as the same message might be handled by multiple sagas.